### PR TITLE
remove op_expr_helper inside gradient_funcs

### PR DESCRIPTION
**Pull request by VertexC**
Clean op_expr_helper with avaibale functional. 
Clean op_traits, replaced with ComposedAttrMap.

https://github.com/Oneflow-Inc/OneTeam/issues/549

- [x] broadcast_like
- [x] bias_add
- [ ] batch_gather | missing functor for UnsortedBatchSegmentSum
- [x] cast
- [x] copy
- [x] concat | add functor for SplitLike
- [ ] ctc_loss | missing functor for CTCLossGrad
- [ ] deconv | missing functor for ConvNd
- [ ] dim_gather | missing functor for DimScatterAddLike
- [ ]  expand | missing functor for ExpandGrad
- [x] flatten
- [ ] layer_norm_grad | missing functor for LayerNorm
- [ ] math_binary_op | missing functor for BinaryXGrad
- [ ] math_unary_op | missing functor for UnaryGrad
- [x] matmul ｜ more test case to add (batch_matmul, broadcast_matmul and their grads) 
- [ ] scalar_pow | missing functor for ScalarPowGrad
- [ ] soft_max | missing functor for SoftmaxGrad
- [ ] sparse_softmax_cross_entropy | missing functor for SparseSoftmaxCrossEntropyGrad
- [x] reshape 
- [x] split_like
- [x] squeeze
- [x] tensor_scalar_binary 
- [ ] tensor_scalar_binary.TensorScalarDiv | missing functor for BroadcastDivGrad, ScalarDivByTensor
- [x] unsqueeze
- [ ] unsample | missing functor for UnsampleGrad

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/deconv.cpp [hunk 1]
```diff
@@ -37,12 +36,12 @@ class DeConvolutionNd : public OpExprGradFunction<DeConvolutionNdCaptureState> {
...
  private:
-  std::shared_ptr<user_op::UserOpConfTrait> op_trait_;
-  std::shared_ptr<std::string> data_format_;
-  std::shared_ptr<std::vector<int32_t>> padding_before_;
-  std::shared_ptr<std::vector<int32_t>> kernel_size_;
-  std::shared_ptr<std::vector<int32_t>> strides_;
-  std::shared_ptr<std::vector<int32_t>> dilation_rate_;
+  AttrMap base_attrs_;
+  std::string data_format_;
+  std::vector<int32_t> padding_before_;
```
Comment (hjchen2) [inline oneflow/core/autograd/gradient_funcs/deconv.cpp:41]: 放到DeConvolutionNdCaptureState里面去

Comment (VertexC) [reply oneflow/core/autograd/gradient_funcs/deconv.cpp:41]: 这个是不是得有了 functor 把op去掉了再放？在init里用op_expr_helper创建op的时候还是需要这些值的

Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/deconv.cpp:41]: 我看了一下，这几个functor都已经有了的。ConvFilterGrad、Conv1d/2d/3d

**Review by hjchen2 — APPROVED** [review]

File: oneflow/core/autograd/gradient_funcs/transpose.cpp [hunk 2]
```diff
@@ -69,7 +65,7 @@ Maybe<void> Transpose::Apply(const TransposeCaptureState* ctx, const TensorTuple
   grad_perm.resize(ctx->perm.size());
   FOR_RANGE(int32_t, i, 0, ctx->perm.size()) { grad_perm.at(ctx->perm.at(i)) = i; }
   JUST(attrs.SetAttr<std::vector<int32_t>>("perm", grad_perm));
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/transpose.cpp:67]: 这一行没有用了

Comment (VertexC) [reply oneflow/core/autograd/gradient_funcs/transpose.cpp:67]: 噢ok 会在 https://github.com/Oneflow-Inc/oneflow/pull/6147 改掉

Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/transpose.cpp:67]: 没事，我在修 BroadcastMatmul 的 PR 中顺手改了吧

**Pull request merged**